src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+  memcpy(symbol_name_as_vector.data(), symbol.c_str(), symbol_name_length);
+  const uint64_t address_symbol_name = address_code + kScratchPadSize;
+  auto result_write_symbol_name =
+      WriteTraceesMemory(pid, address_symbol_name, symbol_name_as_vector);
+  if (result_write_symbol_name.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_symbol_name.error();
+  }
+
+  // We want to do the following in the tracee:
+  // return_value = dlsym(handle, symbol);
+  // The calling convention is to put the parameters in registers rdi and rsi.
+  // So the handle goes to rdi and the address of the symbol name goes to rsi. Then we load the
+  // address of dlsym into rax and do the call. Assembly in Intel syntax (destination first),
+  // machine code on the right:
+
+  // movabs rax, handle               48 b8 handle
+  // mov rdi, rax                     48 89 c7
+  // movabs rax, address_symbol_name  48 b8 address_symbol_name
+  // mov rsi, rax                     48 89 c6
+  // movabs rax, address_dlsym        48 b8 address_dlsym
+  // call rax                         ff d0
+  // int3                             cc
+  MachineCode code;
+  code.AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(absl::bit_cast<uint64_t>(handle))
+      .AppendBytes({0x48, 0x89, 0xc7})
+      .AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_symbol_name)
+      .AppendBytes({0x48, 0x89, 0xc6})
+      .AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_dlsym)
+      .AppendBytes({0xff, 0xd0})
+      .AppendBytes({0xcc});
+  auto result_write_code = WriteTraceesMemory(pid, address_code, code.GetResultAsVector());
+  if (result_write_code.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_code.error();
+  }
+
+  ExecuteOrDie(pid, original_registers, address_code);
+
+  void* return_value = absl::bit_cast<void*>(GetReturnValueOrDie(pid));
+
+  // Cleanup memory and registers.
+  RestoreRegistersOrDie(original_registers);
+  FreeMemoryOrDie(pid, address_code, memory_size);
+  return return_value;
+}
+
+[[nodiscard]] ErrorMessageOr<void> DlcloseInTracee(pid_t pid, void* handle) {
+  // Figure out address of dlclose.
+  OUTCOME_TRY(address_dlclose, FindFunctionAddress(pid, "dlclose", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee.
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, kScratchPadSize));
+
+  // We want to do the following in the tracee:
+  // dlclose(handle);
+  // The calling convention is to put the parameter in registers rdi. Then we load the address of
+  // address_dlclose into rax and do the call. Assembly in Intel syntax (destination first), machine
+  // code on the right:
+
+  // movabs rax, handle               48 b8 handle
+  // mov rdi, rax                     48 8b f8
```

Comment:
Well, again, can you not do `movabs rdi handle`?

Reply:
My understanding is that there is simply no machine code for this. One only can move 64 bit immediates to eax.